### ISSUE TITLE
Confusing error with transitive qualified imports

### ISSUE THREAD

**Issue by brendanzab** [astampoulis/makam#88]
I'm running into this error with qualified imports:

a.makam:
```
foo : type.
foo : foo.
```

b.makam:
```
%import a.

bar : a.foo -> a.foo -> prop.
bar Foo Foo.
```

c.makam
```
%import a.
%import b.

b.bar a.foo Foo ?
```

```
$ /Users/brendan/dev/makam-error/node_modules/.bin/makam c.makam

	Makam, version 0.7.30


!! Error in file ./c.makam, line 4, characters 1-6:
   Term of type b.a.foo -> b.a.foo -> prop,
   whereas expected a type of form a.foo -> A -> prop.
```

Interestingly, removing `%import a.` from `b.makam` results in a successful run with `Foo := a.foo.`, but that means that you can't run/test `b.makam` independently any-more.

**Comment by astampoulis** [comment]
Hi @brendanzab , thank you for the issue. That's indeed unintuitive behavior but (I think) consistent with the simple/ad-hoc notion of namespaces that Makam supports.

Since it's not documented well, the gist of that notion is that:
-  definitions can happen in a namespace block that qualify the relevant constants elsewhere, delineated through `%extend` and `%end`.
- namespace blocks can be nested
- the same namespace can be extended in multiple places
- `%use foo.` works similarly to textual inclusion of the `foo.makam` file in place
- `%import foo.` expands to `%extend foo. %use foo. %end.`
- `%open bar.` allow refering to identifiers in the namespace `bar` without qualification

If it works for your project, could you structure your code like this instead:

a.makam:
```
%extend a.
foo : type.
foo : foo.
%end.
```

b.makam:
```
%use a.

%extend b.
bar : a.foo -> prop.
...
%end.
```

c.makam:
```
%use a.
%use b.
b.bar a.foo Foo ?
```

I think that should still allow you to run/test each file in isolation.

**Comment by brendanzab** [comment]
Oh thanks for the pointers! I think understand the textual inclusion approach a bit better - it seems like a reasonable approach for the time being. I understand a fancier module/namspacing system can be a bit of work to implement! I've started using `%extend namespace. ... %end.` to that end but I'll let you know how I go! 🙂 

For more context I generally try to avoid `%use`-style constructs in languages because they can be hard for readers - ie. it requires a global search of the source and the dependencies to figure out where things come from. This could be alleviated with editor support and [github/semantic](https://github.com/github/semantic/) integration, but that's a whole other chunk of work 😅.

**Comment by brendanzab** [comment]
Wondering if it would be possible to have a way to bring a single identifier into scope from a namespace… for example:

```
%extend semantics.

    value : type.
    neutral : type.
    closure : type.

    eval : list value -> term -> value -> prop.
    readback : int -> value -> term -> prop.

%end.

%extend typing.

    %extend bidirectional.

        %??? semantics (value, type).

    %end.

%end.
```

I kind of try to avoid opening up things because it makes it harder to track down where things are defined. But yeah, this might be pushing up against the limits of the textual inclusion approach 😬

**Comment by astampoulis** [comment]
Hi @brendanzab,

In terms of textual inclusion, it definitely is problematic compared to a proper module system. The decision to go with the present namespacing approach was mostly based on time and effort required at the time (we wanted to showcase the VeriML use case, which requires including an extending a computational language and a logic language with extra constructs, and the present namespacing notion was a very quick way to do that); then I started using it for namespacing in a number of places in the standard library.

The other reason is that designing a module system that works well for Makam is non-trivial. I felt that the standard λProlog/Teyjus approach with split signatures and implementations was not a good fit, especially as Makam is oriented more towards interactive use cases. Also, modules need to be extensible post definition, to keep with the "incremental" definition of types and predicates. ELPI might have a solution, I can read up on that since it's been a while.

In terms of importing only specific constants, it does require quite a bit of restructuring of how the inclusion system works. (We'd have to record the additions that each module defines, which isn't done right now.) There's also the additional issue of what happens with rule definitions -- do we bring everything in scope by default, or should this too should be selective?

I do think that a better module system would be useful though, but not sure what a good path to get there is, and would probably require more effort than the time I have to work on Makam. In general, I try to avoid large changes to the OCaml core of Makam and push more things to be implemented in Makam itself, while trying to find a way to bootstrap eventually. Part of this is that the OCaml core at this point would either need a rewrite or significant effort to be updated to use modern libraries, `dune`, proper testing, and better engineering overall.

**Comment by brendanzab** [comment]
Yeah, that all makes lots of sense, and I appreciate the background! Yeah I can definitely see how the module system might get tricky. The open world nature of Makam is really cool, but sounds challenging to figure out in terms of modules/namespaces.

> There's also the additional issue of what happens with rule definitions -- do we bring everything in scope by default, or should this too should be selective?

If I understand correctly, I really think the default that most languages these days are taking is to strongly encourage selectively bringing things into scope (see Rust and Elm for example). This makes it easier for other people to read and maintain stuff the absence of more powerful structured editing tooling around version control, editing, and online viewing. I definitely find unclear imports to be a significant frustration when trying to understand academic programming language prototypes, which is why I'm rather strongly in favour of it! 😅 

Anyway I hope this didn't come of as too demanding – I really like what's here already, and I can workaround this stuff! I definitely understand you only have so much time to work on this - as somebody with side projects and work commitments I realise this can be a bit to manage. Was hoping more to just document some of my experiences from the perspective of a new user, so that maybe at some point it may help in the future.

Feel free to close this issue if you're not up for working on this right now!

**Comment by astampoulis** [comment]
> If I understand correctly, I really think the default that most languages these days are taking is to strongly encourage selectively bringing things into scope (see Rust and Elm for example). This makes it easier for other people to read and maintain stuff the absence of more powerful structured editing tooling around version control, editing, and online viewing.

I agree and I do like that development. As a bonus, it makes modelling imports through HOAS much nicer :smile: 

I was also referring to a specific aspect that is particularly tricky, where things are a little bit less clear-cut: the visibility of imported rules. This is a similar situation to visibility of `trait impl`s in Rust, type class instances in Haskell, or implicits in Scala/givens in Dotty; that is, Prolog-like "open" definitions in other type systems. In Rust and Haskell, implementations are not named, and their visibility is tied to the trait and the type it's being implemented for. In Scala, visibility can be more controlled by wrapping things in objects too.

In Makam, this is especially tricky for bringing in things selectively: if I am importing a module that includes a rule like `typeof (tuple E1 E2) (product T1 T2)`, when is this brought into scope? For example, what should happen when I select *not* to import `tuple`, but I still import `product`? Of course, we could tie visibility to visibility of a predicate, but that means that we cannot selectively import rules, only constants.

> Anyway I hope this didn't come of as too demanding – I really like what's here already, and I can workaround this stuff! I definitely understand you only have so much time to work on this - as somebody with side projects and work commitments I realise this can be a bit to manage.

Thank you for being understanding, and your comments definitely did not come of as demanding at all. I very much appreciate the feedback, so please keep it coming :smile: It's also really helpful for me in terms of imagining a future version of Makam. From my side, just wanted to provide some context about why it might be hard to make certain changes in the short-term, though I would love to!

**Comment by brendanzab** [comment]
> I was also referring to a specific aspect that is particularly tricky, where things are a little bit less clear-cut: the visibility of imported rules. This is a similar situation to visibility of `trait impl`s in Rust, type class instances in Haskell, or implicits in Scala/givens in Dotty; that is, Prolog-like "open" definitions in other type systems. In Rust and Haskell, implementations are not named, and their visibility is tied to the trait and the type it's being implemented for. In Scala, visibility can be more controlled by wrapping things in objects too.
> 
> In Makam, this is especially tricky for bringing in things selectively: if I am importing a module that includes a rule like `typeof (tuple E1 E2) (product T1 T2)`, when is this brought into scope? For example, what should happen when I select _not_ to import `tuple`, but I still import `product`? Of course, we could tie visibility to visibility of a predicate, but that means that we cannot selectively import rules, only constants.

Oh yeah, we I actually went off on a discussion about this on the @yeslogic internal chat with @mikday, and we did bring up type class resolution as a similar case, but I forgot to report back.

I was also thinking about [Modular Implicits](https://arxiv.org/pdf/1512.01895.pdf), that throw an error if multiple overlapping modules are imported implicitly. Eg. (inventing syntax) this is fine:
```
import implicit int.additiveMonoid

append 1 2
```
but this would be an error:
```
import implicit int.additiveMonoid
import implicit int.multiplicativeMonoid

append 1 2
```
but you could also be explicit:
```
import implicit int.additiveMonoid
import int.multiplicativeMonoid

append {M = int.multiplicativeMonoid} 1 2
```

Naming rules/clauses is an interesting idea. In my ideal world I'd be able to generate an interactive spec from Makam, and be able to say in a doc comment: 'look at `core.typing.bidirectional.synth.conversion` for more information', with a hyperlink. Perhaps that could help with imports… but this all might get annoying in practise. 🤔 

I wonder if it could help to be able to specify predicates as 'sealed' or 'closed'…

**Comment by brendanzab** [comment]
On the topic of documentation I have some [other rough thoughts on what I'd like for a type system/language specification thing here](https://gist.github.com/brendanzab/47fac049f8b40e449db6a9bd60efa57f) - Makam ticks many of the the things I've been looking for on the spec definition side, but does lack some of presentation things I've been looking for. This is less to do with true 'literate' programming, which is more linear in presentation, and more like API docs. May or may not be inspiring for you?

**Comment by astampoulis** [comment]
Thanks for the thoughts @brendanzab! Being able to name rules is indeed something that I'm considering and it could help with selective rule import; it can help both for referencing and also for producing nicer error messages etc. (Still, we would need to consider the ergonomics of the "default" case where you want to import all rules too.)

There's a few things I've been thinking too about the documentation/executable specification aspect. I also took a look at the bidirectional typing gist you've referenced elsewhere; a blog post I'm working on about mode declarations might use a similar example. If you feel like talking about these a bit, feel free to send me an email (antonis DOT stampoulis AT gmail DOT com)!

**Comment by brendanzab** [comment]
Is it possible to open paths? Eg.

```
%extend ast.
    %extend stratified.
        %extend emantics.
        %end.
    %end.
%end.

%extend core_to_stratified.
    %open ast.stratified.semantics.
%end.
```

I'm doing something similar, and get:

```
$ makam test/init --run-tests

	Makam, version 0.7.35


!! Error in file ./test/../src/pass/core_to_stratified.makam:5:15-35:
   Variable stratified.semantics with type prop does not exist.
```

**Comment by astampoulis** [comment]
Hey @brendanzab , that's not possible right now, but I think it should be a quick fix!

**Comment by brendanzab** [comment]
OHHH so cool! You are the best! 😍
